webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/servers/RemoveNodeService.scala

```diff
@@ -327,69 +493,192 @@ class RemoveNodeServiceImpl(
+                        }
+                      }
     } yield {
-      res
+      deleted.flatten
     }
   }
+}
 
-  /**
+
+/*
+ * Close expected reports for node.
+ * Also delete nodes_info for that node.
+ */
+class CloseNodeConfiguration(expectedReportsRepository: UpdateExpectedReportsRepository) extends PostNodeDeleteAction {
+  override def run(nodeId: NodeId, mode: DeleteMode, info: Option[NodeInfo], status: Set[InventoryStatus]): UIO[Unit] = {
+    for {
+      _ <- NodeLoggerPure.Delete.debug(s"  - close expected reports for '${nodeId.value}'")
+      _ <- expectedReportsRepository.closeNodeConfigurationsPure(nodeId).catchAll(err =>
+             NodeLoggerPure.Delete.error(s"Error when closing expected reports for node ${(nodeId, info).name}")
+           ).unit
+      _ <- expectedReportsRepository.deleteNodeInfos(nodeId).catchAll(err => NodeLoggerPure.Delete.error(err.msg))
+    } yield ()
+  }
+}
+// when the node is a policy server, delete directive/rule/group related to it
+class DeletePolicyServerPolicies(policyServerManagement: PolicyServerManagementService) extends PostNodeDeleteAction {
+  override def run(nodeId: NodeId, mode: DeleteMode, info: Option[NodeInfo], status: Set[InventoryStatus]): UIO[Unit] = {
+    // we can avoid to do LDAP requests if we are sure the node wasn't a policy server
+    info.map(_.isPolicyServer) match {
+      case Some(false) =>
+        UIO.unit
+      case _           =>
+        NodeLoggerPure.Delete.debug(s"  - delete relay related policies in LDAP'${nodeId.value}'") *>
+        policyServerManagement.deleteRelaySystemObjectsPure(nodeId).catchAll(err =>
+          NodeLoggerPure.Delete.error(s"Error when deleting system objects (groups, directives, rules) related to relay server for node ${(nodeId, info).name}: ${err.fullMsg}")
+        )
+    }
+  }
+}
+
+// clean up certification key status (only in move mode, not erase)
+class KeyStatus(ldap: LDAPConnectionProvider[RwLDAPConnection], deletedDit: InventoryDit) extends PostNodeDeleteAction {
+  override def run(nodeId: NodeId, mode: DeleteMode, info: Option[NodeInfo], status: Set[InventoryStatus]): UIO[Unit] = {
+    if(mode == DeleteMode.MoveToRemoved) {
+      NodeLoggerPure.Delete.debug(s"  - reset node key certification status for '${nodeId.value}'") *>
+      (for {
+        con <- ldap
+        res <- con.modify(deletedDit.NODES.NODE.dn(nodeId.value), new Modification(ModificationType.REPLACE, A_KEY_STATUS, UndefinedKey.value))
+      } yield () ).catchAll(err =>
+        NodeLoggerPure.Delete.error(s"Error when removing the certification status of node key ${(nodeId, info).name}: ${err.fullMsg}")
+      )
+    } else UIO.unit
+  }
+}
+
+// clean-up cfengine key - only possible if we still have an inventory
+class CleanUpCFKeys extends PostNodeDeleteAction {
+  override def run(nodeId: NodeId, mode: DeleteMode, info: Option[NodeInfo], status: Set[InventoryStatus]): UIO[Unit] = {
+    info match {
+      case Some(i) =>
```

Comment:
this should be done only if the agent type is a cfengine one, not on DSC (even if it wouldn't fail on dsc)